model/dataProvider/SyncDataProviderCollection.php

```diff
@@ -0,0 +1,166 @@
+use oat\taoSync\model\Exception\SyncDataProviderException;
+use Zend\ServiceManager\ServiceLocatorInterface;
+
+class SyncDataProviderCollection extends ConfigurableService
+{
+    const SERVICE_ID = 'taoSync/SyncDataProviderCollection';
+
+    const OPTION_DATA_PROVIDERS = 'taoSync/dataProviders';
+
+    private $dataProviders;
+
+    private $rootDataProviders = [];
+    private $childDataProviders = [];
+
+    /**
+     * @inheritDoc
+     */
+    public function setServiceLocator(ServiceLocatorInterface $serviceLocator = null)
+    {
+        parent::setServiceLocator($serviceLocator);
+        $this->initDataProviders();
+    }
+
+    /**
+     * @param array $params
+     * @return array
+     * @throws SyncDataProviderException
+     */
+    public function getData($params)
+    {
+        return $this->getProvidersData($this->rootDataProviders, $params);
+    }
+
+    /**
+     * @param array $types
+     * @param array $params
+     * @return array
+     * @throws SyncDataProviderException
+     */
+    private function getProvidersData(array $types, array $params)
+    {
+        $result = [];
+
+        foreach ($types as $type) {
+            $dataProvider = $this->getProvider($type);
+            $result[$type] = $dataProvider->getData($params);
+            if (array_key_exists($type, $this->childDataProviders)) {
+                $childParams = array_merge($params, [$type => $result[$type]]);
+                $result = array_merge($result, $this->getProvidersData($this->childDataProviders[$type], $childParams));
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * @param string $type
+     * @return AbstractDataProvider|bool
+     * @throws SyncDataProviderException
+     */
+    public function getProvider($type)
+    {
+        if (!array_key_exists($type, $this->dataProviders)) {
+            throw new SyncDataProviderException(sprintf('data provider %s not found', $type));
+        }
+        return $this->dataProviders[$type];
+    }
+
+    /**
+     * @throws SyncDataProviderException
+     */
+    private function initDataProviders()
+    {
+        if (is_null($this->dataProviders)) {
+            $dataProviders = $this->getOption(self::OPTION_DATA_PROVIDERS);
+
+            if (!is_array($dataProviders)) {
+                throw new SyncDataProviderException('Data providers not set');
+            }
+            $this->dataProviders = [];
+
+            foreach ($dataProviders as $type => $dataProvider) {
+                if ($dataProvider instanceof AbstractDataProvider) {
+                    $this->propagate($dataProvider);
+                    $this->dataProviders[$dataProvider->getType()] = $dataProvider;
+                }
+            }
+            $this->buildDataProvidersTree();
+        }
+    }
+
+    /**
+     * @throws SyncDataProviderException
+     */
+    private function buildDataProvidersTree()
+    {
+        foreach ($this->dataProviders as $type => $dataProvider) {
```

Comment:
All this logic looks a bit over complicated. It's difficult to understand why do we need it at all. We need to discuss and think about possible ways to make it simpler.

Reply:
most part of this logic deleted. now we can set child data providers in config